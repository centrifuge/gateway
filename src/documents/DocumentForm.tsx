import React from 'react';
import { Box, FormField, Grid, ResponsiveContext } from 'grommet';
import { Formik } from 'formik';
import { get } from 'lodash';
import * as Yup from 'yup';
import { Document } from '../common/models/document';
import { AttrTypes, Schema } from '../common/models/schema';
import { SearchSelect } from '@centrifuge/axis-search-select';
import { Contact } from '../common/models/contact';
import { MultipleSelect } from '@centrifuge/axis-multiple-select';
import { Section } from '../components/Section';
import Comments from './Comments';
import Attributes from './Attributes';
import { ViewModeFormContainer } from '../components/ViewModeFormContainer';


// TODO use function components here
type Props = {
  onSubmit?: (document: Document) => void;
  renderHeader: () => JSX.Element;
  contacts: Contact[];
  schemas?: Schema[],
  mode?: 'edit' | 'view' | 'create',
  document: Document
  selectedSchema?: Schema;
};

type State = {
  submitted: boolean,
  columnGap: string,
  sectionGap: string,
  selectedSchema: Schema
}


export class DocumentForm extends React.Component<Props, State> {

  static defaultProps: Props = {
    onSubmit: () => {
      // do nothing
    },
    renderHeader: () => <></>,
    schemas: [],
    mode: 'create',
    document: {
      attributes: {},
      header: {
        read_access: [],
      },
    },
    contacts: [],
  };

  constructor(props) {
    super(props);
    const { selectedSchema } = props;
    this.state = {
      submitted: false,
      columnGap: 'medium',
      sectionGap: 'none',
      selectedSchema,
    };
  }

  onSubmit = (values) => {
    const { selectedSchema } = this.state;

    //

    const payload = {
      ...values,
      attributes: {
        ...values.attributes,
        // add schema as tech field
        '_schema': {
          type: 'string',
          value: selectedSchema.name,
        },
      },
    };

    this.props.onSubmit && this.props.onSubmit(payload);
  };


  generateValidationSchema = (schema: Schema) => {
    // Attributes validation
    let attributes = {};
    let defaultValues = {};
    if (schema) {
      for (let attr of schema.attributes) {
        const path = `${attr.name}`;
        defaultValues[attr.name] = {
          type: attr.type === AttrTypes.PERCENT ? AttrTypes.STRING : attr.type,
          value: '',
        };
        switch (attr.type) {
          case AttrTypes.DECIMAL:
          case AttrTypes.INTEGER:
          case AttrTypes.PERCENT:
            attributes[path] = Yup.object().shape({
              value: Yup.number()
                .required('This field is required')
                .typeError('must be a number'),
            });
            break;
          case AttrTypes.TIMESTAMP:
            attributes[path] = Yup.object().shape({
              value: Yup.date()
                .required('This field is required')
                .typeError('This field is required'),
            });
            break;
          case AttrTypes.BYTES:
            attributes[path] = Yup.object().shape({
              value: Yup.string()
                .matches(/^0x/, 'bytes must start with 0x')
                .required('This field is required'),
            });
            break;
          default:
            attributes[path] = Yup.object().shape({
              value: Yup.string().required('This field is required'),
            });
            break;
        }
      }
      // Add comments field if schema has the option
      if (schema.formFeatures && schema.formFeatures.comments) {
        defaultValues['comments'] = {
          value: '',
          type: AttrTypes.STRING,
        };
      }
    }


    return {
      validationSchema: Yup.object().shape(
        {
          attributes: Yup.object().shape(attributes),
        }),
      defaultValues,
    };
  };


  render() {

    const { submitted, selectedSchema, sectionGap, columnGap } = this.state;
    const { document, mode, contacts, children, renderHeader } = this.props;
    const isViewMode = mode === 'view';
    const isEditMode = mode === 'edit';
    const { validationSchema, defaultValues } = this.generateValidationSchema(selectedSchema);


    // Make sure document has the right form in order not to break the form
    // This should never be the case
    if (!document.attributes) {
      document.attributes = defaultValues;
    } else {
      document.attributes = {
        ...defaultValues,
        ...document.attributes,
      };
    }
    if (!document.header) {
      document.header = {};
    }


    if (!document.header.read_access || !Array.isArray(document.header.read_access)) {
      document.header.read_access = [];
    }

    // Handle cent ids that are not in contacts
    document.header.read_access.forEach(centId => {
      if (!contacts.find(c => c.address!.toLowerCase() === centId.toLowerCase())) {
        contacts.push({
          name: centId,
          address: centId,
        });
      }
    });


    return (
      <ViewModeFormContainer isViewMode={mode === 'view' ? 'view-mode-form':''}  pad={{ bottom: 'xlarge' }}>
        <ResponsiveContext.Consumer>
          {size => {
            return <Formik
              validationSchema={validationSchema}
              initialValues={document}
              validateOnBlur={submitted}
              validateOnChange={submitted}
              onSubmit={(values, { setSubmitting }) => {
                if (!values) return;
                this.onSubmit(values);
                setSubmitting(true);
              }}
            >
              {
                ({
                   values,
                   errors,
                   handleChange,
                   handleSubmit,
                   setFieldValue,
                 }) => (
                  <form
                    onSubmit={event => {
                      this.setState({ submitted: true });
                      handleSubmit(event);
                    }}
                  >
                    <Box gap={sectionGap}>
                      {renderHeader()}

                      {this.renderDetailsSection(
                        values,
                        errors,
                        handleChange,
                        setFieldValue,
                        isViewMode,
                        isEditMode,
                      )}

                      {children}

                      {selectedSchema && <>
                        <Attributes columnGap={columnGap} schema={selectedSchema} isViewMode={isViewMode} size={size}/>
                        {(selectedSchema.formFeatures && selectedSchema.formFeatures.comments) &&
                        <Comments columnGap={columnGap} isViewMode={isViewMode}/>}
                      </>}

                    </Box>
                  </form>
                )
              }
            </Formik>;
          }}
        </ResponsiveContext.Consumer>

      </ViewModeFormContainer>
    );
  }


  renderDetailsSection = (values, errors, handleChange, setFieldValue, isViewMode, isEditMode) => {
    const { selectedSchema, columnGap } = this.state;
    const { contacts, schemas } = this.props;

    return <Section title="Document Details">
      <Grid gap={columnGap}>
        <FormField
          label="Document Schema"
        >
          <SearchSelect
            disabled={isViewMode || isEditMode}
            labelKey={'name'}
            options={schemas}
            value={selectedSchema}
            onChange={(selected) => {
              this.setState({ selectedSchema: selected });
            }}
          />
        </FormField>

        <FormField
          label="Read Access"
        >
          <MultipleSelect
            search={true}
            disabled={isViewMode}
            labelKey={'name'}
            valueKey={'address'}
            options={contacts}
            value={
              get(values, 'header.read_access').map(v => {
                return contacts.find(c => c.address!.toLowerCase() === v.toLowerCase());
              })
            }
            onChange={(selection) => {
              setFieldValue('header.read_access', selection.map(i => i.address));
            }}
          />
        </FormField>
      </Grid>
    </Section>;
  };


};

export default DocumentForm;


