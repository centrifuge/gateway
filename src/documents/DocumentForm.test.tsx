import React from 'react';
import { mount } from 'enzyme';
import DocumentForm from './DocumentForm';
import { BrowserRouter } from 'react-router-dom';

describe('DocumentForm', () => {

  const documents = [
    {
      _id: 'first_id',
      createdAt: '2019-07-09T10:54:59.900Z',
      attributes: {

        ['_schema']: {
          key:
            '0x9ed63b1df0c1b6dc14b777a767ccb0562b7a0adf6f51bf0d90476f6833005f9a',
          type: 'string',
          value: 'first_schema',
        },

        ['reference_id']: {
          key:
            '0x9ed63b1df0c1b6dc14b777a767ccb0562b7a0adf6f51bf0d90476f6833005f9a',
          type: 'string',
          value: 'reference nr1',
        },

        ['customer']: {
          key:
            '0x9ed63b1df0c1b6dc14b777a767ccb0562b7a0adf6f51bf0d90476f6833005f9a',
          type: 'string',
          value: 'some customer',
        },

        ['amount']: {
          key:
            '0x9ed63b1df0c1b6dc14b777a767ccb0562b7a0adf6f51bf0d90476f6833005f9a',
          type: 'decimal',
          value: '100',
        },
      },
    },
    {
      _id: 'second_id',
      createdAt: '2019-07-09T10:54:59.900Z',
      attributes: {

        ['reference_id']: {
          key:
            '0x9ed63b1df0c1b6dc14b777a767ccb0562b7a0adf6f51bf0d90476f6833005f9a',
          type: 'string',
          value: 'reference nr2',
        },

        ['customer']: {
          key:
            '0x9ed63b1df0c1b6dc14b777a767ccb0562b7a0adf6f51bf0d90476f6833005f9a',
          type: 'string',
          value: 'some customer2',
        },

        ['amount']: {
          key:
            '0x9ed63b1df0c1b6dc14b777a767ccb0562b7a0adf6f51bf0d90476f6833005f9a',
          type: 'integer',
          value: '101',
        },
      },
    },
  ];


  const schemas = [
    {
      name: 'first_schema',
      registries:[],
      attributes: [
        {
          name: 'reference_id',
          label: 'Reference Id',
          type: 'string',
        },
        {
          name: 'amount',
          label: 'Amount',
          type: 'decimal',
        },
        {
          name: 'index',
          label: 'Index',
          type: 'integer',
        },
        {
          name: 'date',
          label: 'Some Date',
          type: 'timestamp',
        },
        {
          name: 'customer',
          label: 'Customer',
          type: 'string',
        },
      ],
    },
    {
      name: 'second_schema',
      registries:[],
      attributes: [
        {
          name: 'reference_id',
          label: 'Reference Id',
          type: 'string',
        },
        {
          name: 'amount',
          label: 'Amount',
          type: 'decimal',
        },
        {
          name: 'customer',
          label: 'Customer',
          type: 'string',
        },
      ],
    },
  ];


  const onSubmit = jest.fn(() => {
  });

  const onCancel = jest.fn(() => {
  });

  it('Should render the form just select form', () => {

    const documentForm = mount(
      <DocumentForm document={{}} schemas={schemas} onCancel={onCancel} onSubmit={onSubmit}/>,
    );
    expect(documentForm.html()).toMatchSnapshot();
  });

  it('Should render the form with default data', () => {
    const documentForm = mount(
      <DocumentForm document={documents[0]} schemas={schemas} onCancel={onCancel} onSubmit={onSubmit}/>,
    );
    expect(documentForm.html()).toMatchSnapshot();
  });


  it('Should render the form in view mode', () => {
    const documentForm = mount(
      <DocumentForm document={documents[0]} mode={'view'} schemas={schemas} onCancel={onCancel} onSubmit={onSubmit}/>,
    );
    expect(documentForm.html()).toMatchSnapshot();
  });

  it('Should render the form in edit mode', () => {
    const documentForm = mount(
      <DocumentForm document={documents[0]} mode={'view'} schemas={schemas} onCancel={onCancel} onSubmit={onSubmit}/>,
    );
    expect(documentForm.html()).toMatchSnapshot();
  });


})
;
