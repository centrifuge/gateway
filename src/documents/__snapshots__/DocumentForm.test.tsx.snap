// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentForm Should render the form in edit mode 1`] = `"<div class=\\"StyledBox-sc-13pk1d4-0 hMUzWw\\"><form><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"StyledBox-sc-13pk1d4-0 hroMnY\\"><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Reference Id</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" disabled=\\"\\" name=\\"attributes.reference_id.value\\" class=\\"StyledTextInput-sc-1x30a0s-0 dlXaQi\\" value=\\"reference nr1\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Amount</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div disabled=\\"\\" value=\\"100\\" name=\\"attributes.amount.value\\" class=\\"NumberInput__StyledCurrencyContainer-sc-1mk0e8s-1 fvlnYL\\"><input type=\\"text\\" disabled=\\"\\" name=\\"attributes.amount.value\\" theme=\\"[object Object]\\" class=\\"NumberInput__StyledCurrencyInput-sc-1mk0e8s-0 jqervJ\\" value=\\"100.00\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Index</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div disabled=\\"\\" value=\\"\\" name=\\"attributes.index.value\\" class=\\"NumberInput__StyledCurrencyContainer-sc-1mk0e8s-1 fvlnYL\\"><input type=\\"text\\" disabled=\\"\\" name=\\"attributes.index.value\\" theme=\\"[object Object]\\" class=\\"NumberInput__StyledCurrencyInput-sc-1mk0e8s-0 jqervJ\\" value=\\"0\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Some Date</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><button class=\\"DateInput__StyledDateInputDropButton-sc-1ne1t7m-1 hvzYTL StyledButton-sc-323bzc-0 chnll\\" aria-label=\\"Open Drop\\" disabled=\\"\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 bYNIbh\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" tabindex=\\"-1\\" readonly=\\"\\" class=\\"DateInput__DateTimeTextInput-sc-1ne1t7m-0 dzhNFM StyledTextInput-sc-1x30a0s-0 duvNfq\\" value=\\"\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 ffeAql\\"></div></div></button></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Customer</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" disabled=\\"\\" name=\\"attributes.customer.value\\" class=\\"StyledTextInput-sc-1x30a0s-0 dlXaQi\\" value=\\"some customer\\"></div></div></div></div></div></div></form></div>"`;

exports[`DocumentForm Should render the form in view mode 1`] = `"<div class=\\"StyledBox-sc-13pk1d4-0 hMUzWw\\"><form><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"StyledBox-sc-13pk1d4-0 hroMnY\\"><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Reference Id</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" disabled=\\"\\" name=\\"attributes.reference_id.value\\" class=\\"StyledTextInput-sc-1x30a0s-0 dlXaQi\\" value=\\"reference nr1\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Amount</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div disabled=\\"\\" value=\\"100\\" name=\\"attributes.amount.value\\" class=\\"NumberInput__StyledCurrencyContainer-sc-1mk0e8s-1 fvlnYL\\"><input type=\\"text\\" disabled=\\"\\" name=\\"attributes.amount.value\\" theme=\\"[object Object]\\" class=\\"NumberInput__StyledCurrencyInput-sc-1mk0e8s-0 jqervJ\\" value=\\"100.00\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Index</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div disabled=\\"\\" value=\\"\\" name=\\"attributes.index.value\\" class=\\"NumberInput__StyledCurrencyContainer-sc-1mk0e8s-1 fvlnYL\\"><input type=\\"text\\" disabled=\\"\\" name=\\"attributes.index.value\\" theme=\\"[object Object]\\" class=\\"NumberInput__StyledCurrencyInput-sc-1mk0e8s-0 jqervJ\\" value=\\"0\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Some Date</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><button class=\\"DateInput__StyledDateInputDropButton-sc-1ne1t7m-1 hvzYTL StyledButton-sc-323bzc-0 chnll\\" aria-label=\\"Open Drop\\" disabled=\\"\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 bYNIbh\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" tabindex=\\"-1\\" readonly=\\"\\" class=\\"DateInput__DateTimeTextInput-sc-1ne1t7m-0 dzhNFM StyledTextInput-sc-1x30a0s-0 duvNfq\\" value=\\"\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 ffeAql\\"></div></div></button></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Customer</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" disabled=\\"\\" name=\\"attributes.customer.value\\" class=\\"StyledTextInput-sc-1x30a0s-0 dlXaQi\\" value=\\"some customer\\"></div></div></div></div></div></div></form></div>"`;

exports[`DocumentForm Should render the form just select form 1`] = `"<div class=\\"StyledBox-sc-13pk1d4-0 hMUzWw\\"><form><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"StyledBox-sc-13pk1d4-0 hroMnY\\"><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Document Schema</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><button class=\\"Select__StyledSelectDropButton-sc-17idtfo-1 gQjIFS StyledButton-sc-323bzc-0 jfbKsv\\" aria-label=\\"Open Drop\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 erazuK\\"><div class=\\"StyledBox-sc-13pk1d4-0 XINUX\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" placeholder=\\"Select\\" tabindex=\\"-1\\" type=\\"text\\" readonly=\\"\\" class=\\"Select__SelectTextInput-sc-17idtfo-0 lmyIUy StyledTextInput-sc-1x30a0s-0 jfnKzs\\" value=\\"\\"></div></div><div style=\\"min-width: auto;\\" class=\\"StyledBox-sc-13pk1d4-0 eLQtUu\\"><svg aria-label=\\"FormDown\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 dSllCE\\"><polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" points=\\"18 9 12 15 6 9\\"></polyline></svg></div></div></button></div></div></div></div></div></form></div>"`;

exports[`DocumentForm Should render the form with default data 1`] = `"<div class=\\"StyledBox-sc-13pk1d4-0 hMUzWw\\"><form><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"StyledBox-sc-13pk1d4-0 hroMnY\\"><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Document Schema</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><button class=\\"Select__StyledSelectDropButton-sc-17idtfo-1 gQjIFS StyledButton-sc-323bzc-0 jfbKsv\\" aria-label=\\"Open Drop\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 erazuK\\"><div class=\\"StyledBox-sc-13pk1d4-0 XINUX\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" placeholder=\\"Select\\" tabindex=\\"-1\\" type=\\"text\\" readonly=\\"\\" class=\\"Select__SelectTextInput-sc-17idtfo-0 lmyIUy StyledTextInput-sc-1x30a0s-0 jfnKzs\\" value=\\"first_schema\\"></div></div><div style=\\"min-width: auto;\\" class=\\"StyledBox-sc-13pk1d4-0 eLQtUu\\"><svg aria-label=\\"FormDown\\" viewBox=\\"0 0 24 24\\" class=\\"StyledIcon-ofa7kd-0 dSllCE\\"><polyline fill=\\"none\\" stroke=\\"#000\\" stroke-width=\\"2\\" points=\\"18 9 12 15 6 9\\"></polyline></svg></div></div></button></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 IRSNj\\"><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Reference Id</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" name=\\"attributes.reference_id.value\\" class=\\"StyledTextInput-sc-1x30a0s-0 duvNfq\\" value=\\"reference nr1\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Amount</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div value=\\"100\\" name=\\"attributes.amount.value\\" class=\\"NumberInput__StyledCurrencyContainer-sc-1mk0e8s-1 fvlnYL\\"><input type=\\"text\\" name=\\"attributes.amount.value\\" theme=\\"[object Object]\\" class=\\"NumberInput__StyledCurrencyInput-sc-1mk0e8s-0 eQOnKh\\" value=\\"100.00\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Index</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div value=\\"\\" name=\\"attributes.index.value\\" class=\\"NumberInput__StyledCurrencyContainer-sc-1mk0e8s-1 fvlnYL\\"><input type=\\"text\\" name=\\"attributes.index.value\\" theme=\\"[object Object]\\" class=\\"NumberInput__StyledCurrencyInput-sc-1mk0e8s-0 eQOnKh\\" value=\\"0\\"></div></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Some Date</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><button class=\\"DateInput__StyledDateInputDropButton-sc-1ne1t7m-1 ldQWmt StyledButton-sc-323bzc-0 jfbKsv\\" aria-label=\\"Open Drop\\" type=\\"button\\"><div class=\\"StyledBox-sc-13pk1d4-0 bYNIbh\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" tabindex=\\"-1\\" readonly=\\"\\" class=\\"DateInput__DateTimeTextInput-sc-1ne1t7m-0 dzhNFM StyledTextInput-sc-1x30a0s-0 duvNfq\\" value=\\"\\"></div><div class=\\"StyledBox-sc-13pk1d4-0 ffeAql\\"></div></div></button></div></div><div class=\\"StyledBox__StyledBoxGap-sc-13pk1d4-1 irlfVP\\"></div><div class=\\"FormField__FormFieldBox-m9hood-0 eshLKp StyledBox-sc-13pk1d4-0 CAsMy\\"><label class=\\"StyledText-sc-1sadyjn-0 bKLvLm\\">Customer</label><div class=\\"StyledBox-sc-13pk1d4-0 gZSidW\\"><div class=\\"StyledTextInput__StyledTextInputContainer-sc-1x30a0s-1 kYQCpm\\"><input autocomplete=\\"off\\" name=\\"attributes.customer.value\\" class=\\"StyledTextInput-sc-1x30a0s-0 duvNfq\\" value=\\"some customer\\"></div></div></div></div></div></div></form></div>"`;
